external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala

```diff
@@ -342,80 +415,103 @@ private[kafka010] object CachedKafkaConsumer extends Logging {
-    val groupId = kafkaParams.get(ConsumerConfig.GROUP_ID_CONFIG).asInstanceOf[String]
-    val topicPartition = new TopicPartition(topic, partition)
-    val key = CacheKey(groupId, topicPartition)
-
-    // If this is reattempt at running the task, then invalidate cache and start with
-    // a new consumer
+  def acquire(
+      topicPartition: TopicPartition,
+      kafkaParams: ju.Map[String, Object],
+      useCache: Boolean): KafkaDataConsumer = synchronized {
+    val key = new CacheKey(topicPartition, kafkaParams)
+    val existingInternalConsumer = cache.get(key)
+
+    lazy val newInternalConsumer = new InternalKafkaConsumer(topicPartition, kafkaParams)
+
     if (TaskContext.get != null && TaskContext.get.attemptNumber >= 1) {
-      removeKafkaConsumer(topic, partition, kafkaParams)
-      val consumer = new CachedKafkaConsumer(topicPartition, kafkaParams)
-      consumer.inuse = true
-      cache.put(key, consumer)
-      consumer
-    } else {
-      if (!cache.containsKey(key)) {
-        cache.put(key, new CachedKafkaConsumer(topicPartition, kafkaParams))
+      // If this is reattempt at running the task, then invalidate cached consumer if any and
+      // start with a new one.
+      if (existingInternalConsumer != null) {
+        if (existingInternalConsumer.inuse) {
+          // Consumer exists in cache and is somehow in use. Don't close it immediately, but
+          // mark it for being closed when it is released.
+          existingInternalConsumer.markedForClose = true
+          NonCachedKafkaDataConsumer(newInternalConsumer)
+
+        } else {
+          // Consumer exists in cache and is not in use, so close it immediately and replace
+          // it with a new one.
+          existingInternalConsumer.close()
+          cache.put(key, newInternalConsumer)
+          CachedKafkaDataConsumer(newInternalConsumer)
+
+        }
+      } else {
+        // Consumer is not cached, put the new one in the cache
+        cache.put(key, newInternalConsumer)
+        CachedKafkaDataConsumer(newInternalConsumer)
+
       }
-      val consumer = cache.get(key)
-      consumer.inuse = true
-      consumer
+    } else if (!useCache) {
+      // If planner asks to not reuse consumers, then do not use it, return a new consumer
+      NonCachedKafkaDataConsumer(newInternalConsumer)
+
+    } else if (existingInternalConsumer == null) {
+      // If consumer is not already cached, then put a new in the cache and return it
+      newInternalConsumer.inuse = true
+      cache.put(key, newInternalConsumer)
+      CachedKafkaDataConsumer(newInternalConsumer)
+
+    } else if (existingInternalConsumer.inuse) {
+      // If consumer is already cached but is currently in use, then return a new consumer
+      NonCachedKafkaDataConsumer(newInternalConsumer)
+
+    } else {
+      // If consumer is already cached and is currently not in use, then return that consumer
+      CachedKafkaDataConsumer(existingInternalConsumer)
```

Comment:
we should set `existingInternalConsumer.isuse = true`

Reply:
I wonder why this was not caught in the stress test.

Reply:
i will run a longer stress test locally just to be sure.